### PR TITLE
Fixes #29434 - Add basic documentation for template macros

### PR DESCRIPTION
**Pull request by ofedoren**
As some of you may know we want to enhance our documentation for templates writing. This PR introduces the new tool [apipie-dsl](https://github.com/ofedoren/apipie-dsl) which is very similar to apipie-rails and few examples how to document macros/classes.

Although the PR is WIP, any kind of review as well as help with documentation will be appreciated.

To test this just checkout and run `bundle install`. The documentation will be available under `/templates_doc` url. To speed up documentation rendering you might want to cache it first via `FOREMAN_APIPIE_LANGS=en bundle rake apipie_dsl:cache`.

Also, thanks a lot @ares, @shiramax and @adamruzicka for your initial input!

A link button for documentation is also available in `About` section:
ScreenShot-1585570820456

### CODE REVIEW

File: lib/foreman/renderer/scope/macros/host_template.rb [hunk 1]
```diff
@@ -61,7 +225,7 @@ def enc
           end
 
           def check_host
-            raise HostUnknown if host.nil?
+            raise c if host.nil?
```
Comment (adamruzicka) [inline lib/foreman/renderer/scope/macros/host_template.rb]: Maybe I'm missing something, but what is `c` in this context?

File: app/models/host/base.rb [hunk 2]
```diff
@@ -293,26 +307,48 @@ def overwrite=(value)
...
+      returns Array, 'An array of all managed interfaces'
+      example '@host.managed_interfaces.size # => 3'
+    end
     def managed_interfaces
       interfaces.managed.is_managed.all
     end
 
+    apipie :method, desc: 'Returns an array of all managed bond interfaces objects' do
+      returns Array, 'An array of all bond interfaces'
+      example '@host.bond_interfaces.map { |i| i.ip }.join(",") # => ["192.168.0.1", "10.0.0.1"]'
```
Comment (adamruzicka) [inline app/models/host/base.rb]: This would be `"192.168.0.1,10.0.0.1"`, wouldn't it?

File: app/models/host/managed.rb [hunk 3]
```diff
@@ -650,7 +735,7 @@ def provider
     end
   end
 
-  # no need to store anything in the db if the password is our default
+  #  no need to store anything in the db if the password is our default
```
Comment (adamruzicka) [inline app/models/host/managed.rb]: This shouldn't be needed

**Review by adamruzicka — COMMENTED** [review]
I like this, some nitpicks inline

**Comment by ofedoren** [comment]
Thanks, @adamruzicka! Fixed.

File: app/models/domain.rb [hunk 4]
```diff
@@ -1,6 +1,14 @@
 require "resolv"
 # This models a DNS domain and so represents a site.
 class Domain < ApplicationRecord
+  extend ApipieDSL::Class
```
Comment (tbrisker) [inline app/models/domain.rb]: This is already extended in `ApplicationRecord`, do we need to do this in every class?

**Comment by ofedoren** [comment]
@tbrisker, as per discussion with @ares the PR is changed to contain only basic Ruby methods descriptions and `apipie-dsl` configuration.

In this PR I've added some documentation for `Array`, `Integer` and `Float` classes (we have them mentioned in our current help for template creation).

There are way more classes that we may/want to document (as well as methods in mentioned classes), so I'll add them in separate `refs` PRs so the review process can be easier at least a bit.

Also, @ares, I've removed `nokogiri` dependency in `apipie-dsl` because I've decided to not use it.

**Comment by ofedoren** [comment]
Here's packaging PR: https://github.com/theforeman/foreman-packaging/pull/4970

**Comment by ares** [comment]
I have applied this change on production like system and install the apipie-dsl from the scratch build. When I access `/templates_doc` after foreman reboot, I'm getting 404. When I run `foreman-rake apipie_dsl:cache` it works!

I think it's a good base and can go in. The packaging PR is acked. I'd like to see Foreman tests running, they now fail due to rails upgrade I guess. @ofedoren could you please rebase this on top of development branch? I think we're good to go then.

**Comment by ares** [comment]
[test foreman] seems like unrelated errors, but they are all in postgres-integrations and it's 63 failures

**Review by ares — APPROVED** [review]
Ok, now I see the test fail pretty consistently with other PRs. That should block this (the risk is minimal). @ofedoren I believe we need to get packaging PR in first, please take a look there, I think the definition for el8 is now also needed. However this one can be merged afterwards!

**Comment by tbrisker** [comment]
@ofedoren please add this to the headline features section in the nightly manual https://github.com/theforeman/theforeman.org/blob/gh-pages/_includes/manuals/nightly/1.2_release_notes.md

**Comment by tbrisker** [comment]
Packaging for the new gem has been merged, please open a pr to update the foreman package dependencies as well

**Comment by ofedoren** [comment]
> Packaging for the new gem has been merged, please open a pr to update the foreman package dependencies as well

I don't quite follow where should I do that..

**Comment by tbrisker** [comment]
> > Packaging for the new gem has been merged, please open a pr to update the foreman package dependencies as well
> 
> I don't quite follow where should I do that..

https://github.com/theforeman/foreman-packaging/tree/rpm/develop#foreman - this is needed to make sure the foreman rpm pulls in this new dependency

**Comment by ares** [comment]
@tbrisker besides the dependency and headline feature, is there anything else you had in mind? I'm happy to merge, once done, if not.

**Review by tbrisker — APPROVED** [review]
LGTM other than those comments. Thanks @ofedoren !

**Comment by ofedoren** [comment]
@tbrisker, please see:
https://github.com/theforeman/theforeman.org/pull/1572
https://github.com/theforeman/foreman-packaging/pull/5101

**Comment by ares** [comment]
Seeing the packaging PR updated and the manual too, I'm merging this now. Thanks @ofedoren, looking for more PRs with the documentation! I'll see if I can add a better "help" page content soon.

**Pull request merged**